File: frontend/src/types/paperTrade.ts
export type TradeAction = 'buy' | 'sell';

export type AccountUninitialized = { initialized: false };

export interface AccountInitialized {
  initialized: true;
  initial_cash: number;
  cash_balance: number;
  started_at: string;
  total_value: number;
  return_pct: number;
}

export type AccountResponse = AccountInitialized | AccountUninitialized;

export interface PaperTrade {
  id: number;
  symbol: string;
  action: TradeAction;
  quantity: number;
  price: number;
  total_amount: number;
  realized_pl: number | null;
  executed_at: string;
  note: string | null;
}

export interface TradesPage {
  items: PaperTrade[];
  total: number;
}

export interface PaperHolding {
  id: number;
  symbol: string;
  name: string | null;
  quantity: number;
  avg_price: number;
  current_price: number | null;
  market_value: number | null;
  unrealized_pl: number | null;
  unrealized_pl_pct: number | null;
}

export interface PaperSummary {
  initial_cash: number;
  cash_balance: number;
  holdings_value: number;
  total_value: number;
  unrealized_pl: number;
  realized_pl: number;
  return_pct: number;
  started_at: string;
}

export interface ChartPoint {
  date: string;
  cash: number;
  holdings_value: number;
  total_value: number;
}

export interface PerformanceItem {
  symbol: string;
  name: string | null;
  total_buy_amount: number;
  total_sell_amount: number;
  realized_pl: number;
  unrealized_pl: number;
  total_pl: number;
  return_pct: number | null;
  trade_count: number;
  win_count: number;
}

export interface SummaryMetrics {
  total_pl: number;
  realized_pl: number;
  unrealized_pl: number;
  return_pct: number | null;
  trade_count: number;
  buy_count: number;
  sell_count: number;
  win_count: number;
  loss_count: number;
  win_rate: number | null;
  avg_holding_days: number | null;
  best_trade_pl: number | null;
  worst_trade_pl: number | null;
  profit_factor: number | null;
  expectancy: number | null;
}

export interface PositionCycle {
  entry_date: string;
  exit_date: string;
  entry_price: number;
  exit_price: number;
  quantity: number;
  pl: number;
  return_pct: number;
  holding_days: number;
}

export interface OpenPosition {
  quantity: number;
  avg_price: number;
  current_price: number | null;
  unrealized_pl: number | null;
  unrealized_pl_pct: number | null;
  entry_date: string;
  holding_days: number;
  mfe: number | null;
  mae: number | null;
}

export interface TradeMarker {
  date: string;
  action: TradeAction;
  price: number;
  quantity: number;
}

export interface PricePoint {
  date: string;
  close: number;
}

export interface TimingData {
  price_series: PricePoint[];
  trade_markers: TradeMarker[];
}

export interface BuyAndHold {
  first_buy_date: string | null;
  first_buy_price: number | null;
  bh_value_now: number | null;
  bh_return_pct: number | null;
  actual_return_pct: number | null;
  diff_pct: number | null;
}

export interface EquityPoint {
  date: string;
  invested: number;
  realized_pl: number;
  unrealized_pl: number;
  total_pl: number;
}

export interface AnalyticsResponse {
  symbol: string;
  name: string | null;
  summary: SummaryMetrics;
  position_cycles: PositionCycle[];
  open_position: OpenPosition | null;
  timing: TimingData;
  buy_and_hold: BuyAndHold;
  equity_timeseries: EquityPoint[];
}

export interface TradeCreatePayload {
  action: TradeAction;
  symbol: string;
  quantity: number;
  price?: number;
  executed_at?: string;
  note?: string | null;
  name?: string | null;
}

export interface TradeCreateResponse {
  trade: PaperTrade;
  cash_balance: number;
  total_value: number;
}

File: backend/app/schemas/paper_trade.py
"""PaperTrade Pydantic schemas"""

from datetime import datetime, date
from typing import Optional, Literal, List
from pydantic import BaseModel, Field


# ---------- Account ----------

class AccountUninitialized(BaseModel):
    initialized: Literal[False] = False


class AccountInitialized(BaseModel):
    initialized: Literal[True] = True
    initial_cash: float
    cash_balance: float
    started_at: datetime
    total_value: float
    return_pct: float


class AccountInitRequest(BaseModel):
    initial_cash: float = Field(..., gt=0)


class AccountResetRequest(BaseModel):
    initial_cash: Optional[float] = Field(None, gt=0)


# ---------- Trade ----------

class TradeCreate(BaseModel):
    action: Literal["buy", "sell"]
    symbol: str
    quantity: int = Field(..., gt=0)
    price: Optional[float] = Field(None, gt=0)
    executed_at: Optional[datetime] = None
    note: Optional[str] = None


class TradeResponse(BaseModel):
    id: int
    symbol: str
    action: Literal["buy", "sell"]
    quantity: int
    price: float
    total_amount: float
    realized_pl: Optional[float]
    executed_at: datetime
    note: Optional[str]

    class Config:
        from_attributes = True


class TradesPage(BaseModel):
    items: List[TradeResponse]
    total: int


class TradeCreateResponse(BaseModel):
    trade: TradeResponse
    cash_balance: float
    total_value: float


# ---------- Holding ----------

class HoldingResponse(BaseModel):
    id: int
    symbol: str
    name: Optional[str]
    quantity: int
    avg_price: float
    current_price: Optional[float]
    market_value: Optional[float]
    unrealized_pl: Optional[float]
    unrealized_pl_pct: Optional[float]


# ---------- Summary ----------

class SummaryResponse(BaseModel):
    initial_cash: float
    cash_balance: float
    holdings_value: float
    total_value: float
    unrealized_pl: float
    realized_pl: float
    return_pct: float
    started_at: datetime


# ---------- Chart ----------

class ChartPoint(BaseModel):
    date: date
    cash: float
    holdings_value: float
    total_value: float


# ---------- Performance (symbol list) ----------

class PerformanceItem(BaseModel):
    symbol: str
    name: Optional[str]
    total_buy_amount: float
    total_sell_amount: float
    realized_pl: float
    unrealized_pl: float
    total_pl: float
    return_pct: Optional[float]
    trade_count: int
    win_count: int


# ---------- Symbol Analytics ----------

class SummaryMetrics(BaseModel):
    total_pl: float
    realized_pl: float
    unrealized_pl: float
    return_pct: Optional[float]
    trade_count: int
    buy_count: int
    sell_count: int
    win_count: int
    loss_count: int
    win_rate: Optional[float]
    avg_holding_days: Optional[float]
    best_trade_pl: Optional[float]
    worst_trade_pl: Optional[float]
    profit_factor: Optional[float]
    expectancy: Optional[float]


class PositionCycle(BaseModel):
    entry_date: datetime
    exit_date: datetime
    entry_price: float
    exit_price: float
    quantity: int
    pl: float
    return_pct: float
    holding_days: int


class OpenPosition(BaseModel):
    quantity: int
    avg_price: float
    current_price: Optional[float]
    unrealized_pl: Optional[float]
    unrealized_pl_pct: Optional[float]
    entry_date: datetime
    holding_days: int
    mfe: Optional[float]
    mae: Optional[float]


class TradeMarker(BaseModel):
    date: datetime
    action: Literal["buy", "sell"]
    price: float
    quantity: int


class PricePoint(BaseModel):
    date: date
    close: float


class TimingData(BaseModel):
    price_series: List[PricePoint]
    trade_markers: List[TradeMarker]


class BuyAndHold(BaseModel):
    first_buy_date: Optional[datetime]
    first_buy_price: Optional[float]
    bh_value_now: Optional[float]
    bh_return_pct: Optional[float]
    actual_return_pct: Optional[float]
    diff_pct: Optional[float]


class EquityPoint(BaseModel):
    date: date
    invested: float
    realized_pl: float
    unrealized_pl: float
    total_pl: float


class AnalyticsResponse(BaseModel):
    symbol: str
    name: Optional[str]
    summary: SummaryMetrics
    position_cycles: List[PositionCycle]
    open_position: Optional[OpenPosition]
    timing: TimingData
    buy_and_hold: BuyAndHold
    equity_timeseries: List[EquityPoint]
